resources/analysisTools/qcPipeline/bashLib.sh

```diff
@@ -124,6 +124,216 @@ stringJoin () {
+# * Current pipe extender functions are extendPipe (for single pipes) and extendPipePair (for r1/r2 pipes)
+# * To work with paired pipes (r1/r2) mkPairedPipeName is used to calculate the r1_ or r2_ pipeBaseName
+
+_pipePath="$RODDY_SCRATCH"
+
+# Maintain a mapping of pipe-basenames to current pipe-ends.
+initPipeEnds() {
+    unset _pipeEnds
+    declare -Ag _pipeEnds=()
+}
+
+listPipeEnds() {
+    for i in "${!_pipeEnds[@]}"; do
+        echo "'$i' => '${_pipeEnds[$i]}'"
+    done
+}
+
+# Make a path without registering it.
+mkPipePath() {
+    local pipeName="${1:?Named-pipe base name}"
+    echo "$_pipePath/$pipeName"
+}
+
+# Just get the path from the registry.
+getPipeEndPath() {
+    local pipeName="${1:?No pipename}"
+    if [[ ! ${_pipeEnds[$pipeName]+_} ]]; then
+        throw 152 "Could not find pipe by name '$pipeName'"
+    fi
+    local pipePath="${_pipeEnds[$pipeName]}"
+    echo "$pipePath"
+}
+
+# Register a path in the registry under the key "pipeName".
+setPipeEndPath() {
+    local pipeName="${1:?No pipename}"
+    local pipePath="${2:?No pipe path}"
+    if [[ -z "$pipeName" ]]; then
+        throw 153 "Cannot set empty pipename"
+    fi
+    _pipeEnds[$pipeName]="$pipePath"
+}
+
+# Create a path from the name and the tag and register it as new pipe-end in the registry.
+# Note that the pipe is registered as tempfile and will be deleted upon cleanUp.
+updatePipeEndPath() {
+    local pipeName="${1:?Named-pipe base name}"
+    local tag="${2:?Pipe tag}"
+    local pipePath=$(mkPipePath "${pipeName}_$tag")
+    mkfifo "$pipePath" || throw 150 "Could not create named pipe at '$pipePath'"
+    registerTmpFile "$pipePath"
+    setPipeEndPath "$pipeName" "$pipePath"
+}
+
+# Create a pipe and register it as new pipe-end in the registry. It will be the source of pipe.
+mkPipeSource() {
+    local pipeName="${1:?No pipename}"
+    if [[ ${_pipeEnds[$pipeName]+_} ]]; then
+        throw 154 "Cannot create new pipe source. Pipe named '$pipeName' -- already exists with value  '${_pipeEnds[$pipeName]}'"
+    fi
+    updatePipeEndPath "$pipeName" "source"
+}
+
+# Linear extension of a pipeline.
+# For a command with the interface "command infile outfile @rest" take the pipe with the given basename
+# and use it as input file. Create a new output pipe and set that pipe as output file.
+# Set the new end of the linear pipeline to the output pipe. Call like this:
+#
+#   extendPipe $pipeName $tag -- commandName @restArgs
+#
+# The "--" is optional.
+extendPipe() {
+    local pipeName="${1:?No pipe basename}"
+    local tag="${2:?No tag}"
+    shift 2
+    if [[ "$1" == "--" ]]; then
+        shift
+    fi
+    local command="${1:?No command/function}"
+    shift
+    declare -a args=("$@")
+
+    local inpipe=$(getPipeEndPath "$pipeName")
+    updatePipeEndPath "$pipeName" "$tag"
+    local outpipe=$(getPipeEndPath "$pipeName")
+
+    "$command" "$inpipe" "$outpipe" "${args[@]}" & registerPid
```

Comment:
Does refer command to a binary or to a function? I don't see, how this works for all binaries.

Reply:
Currently, a binary or bash function.

Maybe it is possible to say
```bash
$command "$inpipe" ...
```
to allow for a "binary -arg1 -arg2" argument as command that is then interpreted as command prefix. I don't know and could try later.

Reply:
Even though this is rare, it is possible that the path to command contains a space. Without the quotes that would not resolve correctly. I rather request from the user to have the desired interface "command inpipe outpipe", even if he will have to define a function to achieve this.